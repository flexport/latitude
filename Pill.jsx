/**
 * TEAM: frontend_infra
 * @flow strict
 */

import * as React from "react";
import {StyleSheet, css} from "aphrodite";
import Text from "./Text";
import IconButton from "./button/IconButton";
import Link from "./Link";

import latitudeColors from "./colors";

const sizes = {
  xs: 20,
  s: 24,
  m: 30,
  l: 40,
};

const spacing = {
  xs: 7,
  s: 7,
  m: 8,
  l: 12,
};

type Props = {|
  /** The information to display */
  +children: string,
  /** The url the pill redirects to when the text is clicked. If present,
   * the text inside the pill is a link. If absent the text is not a link
   * and has no on click behavior. Links will be opened in a new tab.
   */
  +href?: string,
  /** The function invoked when the Pill is clicked. If no onDismiss is
   * specified, then the Pill will not be closeable and the close button
   * will be hidden */
  +onDismiss?: () => void,
  /** The size of the pill */
  +size?: $Keys<typeof sizes>,
  /** Called when the cursor enters the Pill */
  +onMouseEnter?: (e: Event) => void,
  /** Called when the cursor leaves the Pill */
  +onMouseLeave?: (e: Event) => void,
  /** Called when the Pill is focused */
  +onFocus?: (e: Event) => void,
  /** Whether or not it's possible to highlight text */
  +selectable?: boolean,
|};

/**
 * @short A compact display component for dismissible information
 * @brandStatus V2
 * @status Beta
 * @category Data Display
 *
 * Pills are useful for organizing information like a group of H.S. codes
 * or product tags. Using pills to organize information can provide a
 * straightforward layout and give the user a simple ability to remove
 * one or multiple items.
 */
export default function Pill({
  children,
  href,
  onDismiss,
  size = "m",
  onMouseEnter,
  onMouseLeave,
  onFocus,
  selectable = true,
}: Props) {
  const style = {
    height: sizes[size],
    padding: `0 ${spacing[size]}px`,
    borderRadius: sizes[size] / 2,
    ...(!selectable ? {userSelect: "none"} : null),
  };

  return (
    <span
      className={css(styles.pill)}
      style={style}
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
      onFocus={onFocus}
    >
      <Text
        scale="subtext"
        overflow="hidden"
        textOverflow="ellipsis"
        color="grey40"
        whiteSpace="nowrap"
      >
        {href === undefined || href.length === 0 ? (
          children
        ) : (
          <Link href={href} openInNewTab={true} key={`${children}_link`}>
            {children}
          </Link>
        )}
      </Text>
      {onDismiss ? (
        <span style={{paddingLeft: "12px"}}>
          <IconButton
            type="button"
            kind="blank"
            iconName="cancel"
            onClick={onDismiss}
            height={{type: "customDontUse", height: 12}}
          />
        </span>
      ) : null}
    </span>
  );
}

const styles = StyleSheet.create({
  pill: {
    boxSizing: "border-box",
    display: "inline-flex",
    position: "relative",
    alignItems: "center",
    border: `solid 1px ${latitudeColors.grey20}`,
    maxWidth: "100%",
    background: latitudeColors.white,
    overflow: "hidden",
  },
});
