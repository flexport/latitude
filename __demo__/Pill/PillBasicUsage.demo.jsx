/**
 * TEAM: frontend_infra
 * @flow
 */

import React from "react";

import Pill from "../../Pill";

/**
 * @title Basic Usage
 */
export default function PillBasicUsage() {
  return <Pill onDismiss={() => {}}>ETD: Jan 01, 2019 CST</Pill>;
}
