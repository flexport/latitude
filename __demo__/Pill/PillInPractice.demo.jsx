/**
 * TEAM: frontend_infra
 * @flow
 */

import React, {useState} from "react";
import {Form, Field} from "formula-one";

import Pill from "../../Pill";
import Label from "../../Label";
import Text from "../../Text";
import TextInput from "../../TextInput";
import Button from "../../button/Button";
import DeprecatedHorizontalGroup from "../../DeprecatedHorizontalGroup";
import DeprecatedVerticalGroup from "../../DeprecatedVerticalGroup";

/**
 * @title In-practice example
 */
export default function PillInPractice() {
  const [pills, setPills] = useState([]);

  const handleAdd = (value: string) => {
    setPills([...pills, value]);
  };

  const handleRemove = (index: number) => {
    const newPills = [...pills];
    newPills.splice(index, 1);
    setPills(newPills);
  };

  return (
    <Form initialValue="" onSubmit={handleAdd}>
      {(link, onSubmit) => (
        <DeprecatedVerticalGroup spacing="l">
          <DeprecatedHorizontalGroup crossAlign="end">
            <Field link={link}>
              {(value, _errors, onChange, onBlur) => (
                <Label value="Add a tag">
                  <TextInput
                    value={value}
                    onChange={onChange}
                    onBlur={onBlur}
                  />
                </Label>
              )}
            </Field>
            <Button onClick={onSubmit}>Submit</Button>
          </DeprecatedHorizontalGroup>
          <DeprecatedHorizontalGroup>
            {pills.length > 0 ? (
              pills.map((pill, i) => (
                <Pill key={pill} onDismiss={() => handleRemove(i)}>
                  {pill}
                </Pill>
              ))
            ) : (
              <Text>No tags added.</Text>
            )}
          </DeprecatedHorizontalGroup>
        </DeprecatedVerticalGroup>
      )}
    </Form>
  );
}
