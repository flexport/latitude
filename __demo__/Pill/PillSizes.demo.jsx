/**
 * TEAM: frontend_infra
 * @flow
 */

import React from "react";

import Group from "../../Group";
import Label from "../../Label";
import Pill from "../../Pill";

/**
 * @title Sizes
 */
export default function PillSizes() {
  return (
    <Group flexDirection="column">
      <Label value="Extra Small" />
      <Pill size="xs" onDismiss={() => {}}>
        ETD: Jan 01, 2019 CST
      </Pill>
      <Label value="Small" />
      <Pill size="s" onDismiss={() => {}}>
        ETD: Jan 01, 2019 CST
      </Pill>
      <Label value="Medium" />
      <Pill size="m" onDismiss={() => {}}>
        ETD: Jan 01, 2019 CST
      </Pill>
      <Label value="Large" />
      <Pill size="l" onDismiss={() => {}}>
        ETD: Jan 01, 2019 CST
      </Pill>
    </Group>
  );
}
