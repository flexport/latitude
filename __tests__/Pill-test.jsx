/**
 * TEAM: frontend_infra
 * @flow
 */

import * as React from "react";
import {mount} from "enzyme";
import Pill from "../Pill";

describe("Pill", () => {
  it("works as expected, shows close button if onDismiss is provided", () => {
    expect(mount(<Pill onDismiss={() => {}}>child</Pill>)).toMatchSnapshot();
  });

  it("works as expected, does not show close button if onDismiss is not provided", () => {
    expect(mount(<Pill>child</Pill>)).toMatchSnapshot();
  });

  it("handles onDismiss", () => {
    let clicked = false;
    const handleDismiss = () => {
      clicked = true;
    };
    const pill = mount(<Pill onDismiss={handleDismiss}>pill</Pill>);

    pill.find("button").simulate("click");

    expect(clicked).toBe(true);
  });
});
