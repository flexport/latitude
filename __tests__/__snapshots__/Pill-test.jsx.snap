// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pill works as expected, does not show close button if onDismiss is not provided 1`] = `
<Pill>
  <span
    className="pill_oagavu"
    style={
      Object {
        "borderRadius": 15,
        "height": 30,
        "padding": "0 8px",
      }
    }
  >
    <Text
      color="grey40"
      overflow="hidden"
      scale="subtext"
      textOverflow="ellipsis"
      whiteSpace="nowrap"
    >
      <span
        className="standardOverride_eiiwmj"
        style={
          Object {
            "color": "#67768D",
            "display": "inline-block",
            "fontFamily": Array [
              "gt-america",
              "-apple-system",
              "BlinkMacSystemFont",
              "Segoe UI",
              "Roboto",
              "Oxygen",
              "Ubuntu",
              "Cantarell",
              "Fira Sans",
              "Droid Sans",
              "Helvetica Neue",
              "sans-serif",
            ],
            "fontSize": "12px",
            "fontStyle": "normal",
            "fontWeight": "400",
            "letterSpacing": 0,
            "lineHeight": "16px",
            "overflow": "hidden",
            "textOverflow": "ellipsis",
            "textTransform": "initial",
            "whiteSpace": "nowrap",
            "wordBreak": "normal",
          }
        }
      >
        child
      </span>
    </Text>
  </span>
</Pill>
`;

exports[`Pill works as expected, shows close button if onDismiss is provided 1`] = `
<Pill
  onDismiss={[Function]}
>
  <span
    className="pill_oagavu"
    style={
      Object {
        "borderRadius": 15,
        "height": 30,
        "padding": "0 8px",
      }
    }
  >
    <Text
      color="grey40"
      overflow="hidden"
      scale="subtext"
      textOverflow="ellipsis"
      whiteSpace="nowrap"
    >
      <span
        className="standardOverride_eiiwmj"
        style={
          Object {
            "color": "#67768D",
            "display": "inline-block",
            "fontFamily": Array [
              "gt-america",
              "-apple-system",
              "BlinkMacSystemFont",
              "Segoe UI",
              "Roboto",
              "Oxygen",
              "Ubuntu",
              "Cantarell",
              "Fira Sans",
              "Droid Sans",
              "Helvetica Neue",
              "sans-serif",
            ],
            "fontSize": "12px",
            "fontStyle": "normal",
            "fontWeight": "400",
            "letterSpacing": 0,
            "lineHeight": "16px",
            "overflow": "hidden",
            "textOverflow": "ellipsis",
            "textTransform": "initial",
            "whiteSpace": "nowrap",
            "wordBreak": "normal",
          }
        }
      >
        child
      </span>
    </Text>
    <span
      style={
        Object {
          "paddingLeft": "12px",
        }
      }
    >
      <IconButton
        disabled={false}
        height={
          Object {
            "height": 12,
            "type": "customDontUse",
          }
        }
        iconAlignment="left"
        iconName="cancel"
        intent="none"
        isLoading={false}
        kind="blank"
        onClick={[Function]}
        size="m"
        type="button"
        width="responsive"
      >
        <button
          className="base_i7qsly-o_O-blank_1tdfl6w-o_O-responsive_zk4wup-o_O-button_1cblu7c-o_O-none_v41adi"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          type="button"
        >
          <Icon
            alignment="center"
            deprecatedAllowColorInheritance={true}
            iconName="cancel"
            isARIAAccessible={false}
            size="xs"
          >
            <span
              className="wrapper_17y0hv9 svgIcon"
              style={
                Object {
                  "fill": "unset",
                  "height": 12,
                  "width": 12,
                }
              }
            >
              <svg
                aria-hidden="true"
                aria-labelledby="title desc"
                className="svgCenter_1r81nla"
                color="unset"
                height={12}
                role="img"
                style={
                  Object {
                    "height": "unset",
                    "width": "unset",
                  }
                }
                viewBox="0 0 64 64"
                width={12}
              >
                <title
                  id="title"
                />
                <desc
                  id="desc"
                >
                  X
                </desc>
                <g
                  fillRule="nonzero"
                >
                  <path
                    d="M53.213 46.142L39.071 32l14.142-14.142a1 1 0 0 0 0-1.414l-5.657-5.657a1 1 0 0 0-1.414 0L32 24.929 17.858 10.787a1 1 0 0 0-1.414 0l-5.657 5.657a1 1 0 0 0 0 1.414L24.929 32 10.787 46.142a1 1 0 0 0 0 1.414l5.657 5.657a1 1 0 0 0 1.414 0L32 39.071l14.142 14.142a1 1 0 0 0 1.414 0l5.657-5.657a1 1 0 0 0 0-1.414z"
                    fillRule="evenodd"
                    key="0"
                    role="presentation"
                  />
                </g>
              </svg>
            </span>
          </Icon>
        </button>
      </IconButton>
    </span>
  </span>
</Pill>
`;
